patterns/interpreter.md

```diff
@@ -0,0 +1,185 @@
+
+```ignore
+S -> SS -> ()(S) -> ()()
+```
+
+## Context Free Grammar for our problem
+
+Recall that our task is translate infix expressions into postfix ones. Let's
+define a context free grammar for a set of infix expressions over `0,...,9, +,-`,
+where:
+
+- terminal symbols: `0,...,9, +,-`
+- non-terminal symbols: `exp, term`
+- start symbol is `exp`
+- and the following are production rules
+
+```ignore
+exp -> exp + term
+exp -> exp - term
+exp -> term
+term -> 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9
+```
+
+In fact this grammar should be further transformed depending on what we are going
+to do with the grammar. For example, we might need to remove left recursion.
+For more details please see [Compilers:Principles,Techniques, and Tools](https://en.wikipedia.org/wiki/Compilers:_Principles,_Techniques,_and_Tools)
+(aka Dragon Book).
+
+## Solution
+
+We simply implement a recursive descent parser. The code panics when an expression
+is syntactically wrong (for exmple `2-34` or `2+5-` are wrong according to the
+grammar definition).
+
+```rust
+pub struct Interpreter<'a> {
+    it: std::str::Chars<'a>,
+}
+
+impl<'a> Interpreter<'a> {
+
+    pub fn new(infix: &'a String) -> Self {
+        Self { it: infix.chars() }
+    }
+
+    pub fn interpret(&mut self, out: &mut String) {
+        self.exp(out);
+    }
+
+    fn next_char(&mut self) -> Option<char> {
+        self.it.next()
+    }
+
+    fn exp(&mut self, out: &mut String) {
+        self.term(out);
+
+        while let Some(op) = self.next_char() {
+            if op == '+' || op == '-' {
+                self.term(out);
+                out.push(op);
+            } else {
+                panic!("Unexpected symbol '{}'", op);
+            }
+        }
+    }
+
+    fn term(&mut self, out: &mut String) {
+        match self.next_char() {
+            Some(ch) if ch.is_digit(10) => out.push(ch),
+            Some(ch) => panic!("Unexpected symbol '{}'", ch),
+            None => panic!("Unexpected end of string"),
+        }
+    }
+}
+
+pub fn main() {
+    let infix = String::from("2+3");
+    let mut intr = Interpreter::new(&infix);
+    let mut postfix = String::new();
+    intr.interpret(&mut postfix);
+    assert_eq!(postfix, "23+");
+
+    let infix = String::from("1-2+3-4");
+    intr = Interpreter::new(&infix);
+    postfix.clear();
+    intr.interpret(&mut postfix);
+    assert_eq!(postfix, "12-3+4-");
+}
+```
+
+## Discussion
+
+There may be a wrong perception that the Interpreter design pattern
+is about design grammars for formal languages and implementation
+of parsers for these grammars. In fact, this pattern is about expressing
+problem instances in a more specific way and implementing
+functions/classes/structs solving these problem instances.
+Rust language has `macro_rules!` that allows to define special
+syntax and rules how to expand this syntax into source code.
+
+In the following example we create a simple `macro_rules!` that computes
```

Comment:
is this example part of the Interpreter design pattern? Or it's just an example in which you explain macro rules?

Reply:
@MarcoIeni Using `macro_rules!` is one way to implement interpreter pattern. I am not explaining here how to use `macro_rules!`. The basic idea is that you express a problem in a special syntax like `norm!(...)` rather `let v = vec![...]; norm(&a);...`.

Reply:
Ok.
In the previous version of this article you used macro rules to rewrite the same example you first wrote in code.
This was really effective in my opinion because you clearly showed the power of macro rules.
I wonder if you think this would be useful also here, instead of doing a different example.
I would like to hear also other reviewer's opinion.

Reply:
@MarcoIeni Ok, let's see what others think. I could rewrite the example if we come to a consensus.

Reply:
For me either way is nice, if you both think it's nice to use the same "problem" to show the power of macros, why not. :)

Reply:
it's up to you fade2black :)

Reply:
I just left as is.